### ISSUE TITLE
FunctionList regenerates when using find result to jump to a match

### ISSUE THREAD

**Issue by jmarrec** [notepad-plus-plus/notepad-plus-plus#583]
Hi,

I've used FunctionList in order to create a navigation pane for long text-based building energy modeling input files.

As pointed out by one of my users [here](https://github.com/jmarrec/notepad/issues/1#issuecomment-125128203), when you use Search > Find All  and then click on one of the matches that appear in the "Find result" panel, it will regenerate the function list each time, potentially causing quite a slowdown with huge text files and/or a slow-ish computer.

![Search and replace](https://cloud.githubusercontent.com/assets/10520003/8803925/e65408be-2fc9-11e5-9ea2-ca4c84b864a3.PNG)

**Do you know if there's any way to disable this behavior?**

If not, a workaround would be to disable the function list pane as long when trying to jump through the file by clicking on the hits repeatedly to avoid the freeze.
Is there a way to have a keyboard shortcut to show/hide the functionlist panel? I've looked at the Modify Shortcut/Delete Command panel but couldn't find it in the list.

Thank you!

**Comment by milipili** [comment]
It needs confirmation but for me it seems to be  an invalid behavior from n++. Would it possible to privide a file to easily reproduce the issue ? (A dummy one would be fine)

**Comment by Rikk** [comment]
>then click on one of the matches that appear in the "Find result" panel, it will regenerate the function list each time

I see it too, with small files it just blinks Function List panel, what means a refresh(reload) was triggered. In bigger files the reload takes more time, and only after that the line is selected.

**Comment by jmarrec** [comment]
@milipili: I just created a dummy python file (since syntax highlighting for python ships directly with n++).
```python
 def dummy():

    print("dummy")


```

Repeat until you reach tens of thousand of lines, you won't be able to miss the fact that it reloads each time. Here's a [pastebin](http://pastebin.com/T0HH4iY2#) with 45000 lines for convenience. Reloads takes a little under a second on my machine

**Comment by ariccio** [comment]
> I see it too, with small files it just blinks Function List panel, what means a refresh(reload) was triggered. 

Could someone link to the spot in the source where the reload is triggered?

**Comment by Rikk** [comment]
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/WinControls/FunctionList/functionListPanel.cpp#L206
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/WinControls/FunctionList/functionListPanel.cpp#L422
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/WinControls/FunctionList/functionListPanel.cpp#L666
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/Notepad_plus.cpp#L4740
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/Notepad_plus.cpp#L5375
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/NppIO.cpp#L483
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/NppIO.cpp#L1559

**Comment by ariccio** [comment]
Thanks, @Rikk.

When I get to my computer, I think I'll profile it so I can better understand what's going on.

My best guess is that the easiest "fix" is to just rewrite `FunctionListPanel::reload` to be ***really*** fast, which would be much simpler than some sort of double buffering.

(will elaborate in next comment)

**Comment by Rikk** [comment]
Yea, make it faster should be great... But why it needs to reload when double click entry in 'find result'? File is not being changed, thus no change functions.

**Comment by ariccio** [comment]
I did something similar in another program (`altWinDirStat`), where an MFC `ListView` was flickering.

There were several columns of text, and formatting text was the main time-consumer. Instead of implementing double/triple buffering, I rewrote the string formatting functions to:

- Preallocate memory, as a fixed-size buffer on the stack, enough for *most* strings that I'd need to format.
- Write *directly into the preallocated buffer*, no copying to/from temporaries.
- Keep track of the number of characters written, at each step - to later enable a `DrawText` call with a known length (no need to search for a null-terminator!).
- Extensively use static analysis, so I didn't screw up!

If I remember correctly, I had an order of magnitude improvement, and it scaled very well, for essentially no design changes.

**Comment by ariccio** [comment]
> But why it needs to reload when double click entry in 'find result'? File is not being changed, thus no change in functions.

I agree, there's no good reason to repeat all the work, and it might be a really simple fix. However:

- We should be sure that the file hasn't actually changed, and carefully review the code to verify.
- I suspect that double-clicking the entry in "find result" isn't the only manifestation of slowness in `reload`.

**Comment by ariccio** [comment]
Does anybody know: For the function list, does `SciLexer.dll` need to be compiled with Boost's PCRE?

**Comment by MAPJe71** [comment]
When you want to use PCRE for the parsers in FunctionList.xml: Yes

**Comment by ariccio** [comment]
Grr. I don't have symbols for the boost-enabled `SciLexer.dll`, but *initial* profiling points at `ScintillaWin::CaseMapString`, and `::LCMapStringW` therein, as major time consuming functions.

**Comment by laxamar** [comment]
Just to add my 2c, I often program over a slow VPN and the issue is extremely debilitating in that case. It also seems that once I enable the function pane, closing n++ DOS NOT fix the issue. Is there a way to reset the function list pane to unused?

**Comment by jmarrec** [comment]
Salut Jacques,

Can't you just close the function pane in that case when you're over VPN?
(Or download the file to your local machine but that's probably something you don't want to do)

Julien


Envoyé de mon iPhone

**Comment by laxamar** [comment]
Bonjour Julien,

The problem is that *once* I opened the function list pane, the behavior 
started and never went away, even when I closed the function pane, 
restarted Notepad++,  the behavior remained.  I looks like once the pane 
is activated, the software will always try to refresh even with the 
function list pane closed afterwards. I can easily spend 60 seconds 
looking at the window flash (that's another annoying feature while 
building the list). Which makes any window switch counter productive. I 
have reverted to using direct Linux text edition through ssh :-) (old 
school).

I'm working on a rather large project with thousands of files, some of 
which are very large objects.

Just to be completely thorough, this is the first time I've tried this 
VPN - Notepad++ on my upgraded Windows 10 machine, Before that, my dev 
machine was on Windows 7

Jacques

**Comment by laxamar** [comment]
Just one more note about performance through VPN:

If I disable "File Status Auto-Detection", then system is MUCH more 
responsive ( at the risk of file change collisions!).  It *looks* like 
every tab is checked every time Notepad++ window is selected.
The preferred behavior might be to only check the currently active tab.

Jacques

**Comment by jmarrec** [comment]
@donho It's old, but it's not outdated.

I just tested by reopening the pastebin dummy.py (see [here](https://pastebin.com/T0HH4iY2#)) into Notepad++ v7.6.6 64bit, and I still see the FunctionList taking a long time to recalculate everything I click on an entry in the "Find Result" window (after a "Find All")
